baremetal: Add more documentation about networking.

russellb opened the pull request.

russellb committed:
baremetal: Add more documentation about internal DNS.

stbenjam commented on data/data/bootstrap/baremetal/README.md:
```diff
@@ -22,19 +22,33 @@ Relevant files:
   This runs `keepalived.sh` to generate the proper configuration from the
   template and then runs podman to launch `keepalived`.
+* **files/usr/local/bin/fletcher8** - 
+* **files/usr/local/bin/get_vip_subnet_cidr** - 
```
Should these get descriptions? There are some already in the file, so wasn't sure:

fletcher8:

```
# Script that uses the fletcher8 algorithm to generate a hash from an input
# string. This is used to generate VRRP ids for use with Keepalived.
```

get_vip_subnet_cidr:

```
# Script to determine the network CIDR for a given VIP.
```

russellb replied:
Sure, I'll copy those descriptions

russellb committed:
baremetal: Add descriptions of fletcher8 and get_vip_subnet_cidr.

hardys commented on data/data/bootstrap/baremetal/README.md:
```diff
@@ -22,19 +22,33 @@ Relevant files:
-* files/usr/local/bin/coredns.sh
-* systemd/units/coredns.service
+* **files/etc/keepalived/keepalived.conf.tmpl** - `keepalived` configuration
+  template, includes configuration for the DNS VIP
```
Should we add a note here about the filename extension, e.g it's not .template because the templating is done via envsubst in https://github.com/openshift-metal3/kni-installer/blob/master/data/data/bootstrap/baremetal/files/usr/local/bin/keepalived.sh and not via the typical go templating inside the installer?

russellb replied:
added a note

hardys commented on docs/design/baremetal/networking-infrastructure.md:
```diff
@@ -70,7 +70,101 @@ DNS requirements internal to the cluster as possible, leaving only a
+
+After the bootstrap process completes, the DNS VIP moves to one of the control
+plane nodes.  All control plane nodes will continue to use this DNS VIP as
+their primary DNS server.
```
Should we mention how that's done, e.g via a NM prepender script, or is that documented in the MCO integration?

markmc replied:
++ need to cover dns-vip-prepender

russellb replied:
Good point.  There's no documentation of the MCO integration, though.  I envision this document being the high level docs that discuss how all the pieces together.  Detailed documentation of the specific MCO pieces should perhaps be in the MCO repository.  That's where I'd expect to document the exact piece that performs this task.

Once the MCO docs exist, we should be linking to them from this document.

What do you think?

markmc commented on data/data/bootstrap/baremetal/README.md:
```diff
@@ -22,19 +22,37 @@ Relevant files:
   template and then runs podman to launch `keepalived`.
+* **files/usr/local/bin/fletcher8** - Script that uses the
+  [fletcher8](https://en.wikipedia.org/wiki/Fletcher%27s_checksum) algorithm to
+  generate a hash from an input string. This is used to generate VRRP ids for
```
based on the cluster name, in order to avoid collisions with other clusters on the same network?

russellb replied:
nice one, I've incorporated this

markmc commented on data/data/bootstrap/baremetal/README.md:
```diff
@@ -22,19 +22,37 @@ Relevant files:
+  generate a hash from an input string. This is used to generate VRRP ids for
+  use with Keepalived.  This is used by `keepalived.sh`
+* **files/usr/local/bin/get_vip_subnet_cidr** - Script to determine the network
+  CIDR for a given VIP.  This is used by `keepalived.sh`
```
which is in turn used to determine which local interface is on the VIP's network

russellb replied:
added, thanks!

markmc commented on docs/design/baremetal/networking-infrastructure.md:
```diff
@@ -70,7 +70,101 @@ DNS requirements internal to the cluster as possible, leaving only a
+
+The `mdns-publisher` is the component that runs on each host to make itself
+discoverable by other hosts in the cluster.  It will be run by the
+`machine-config-operator`, once the following PR is complete:
```
it is not run on the bootstrap node by the installer because ...

russellb replied:
Added:

```
`mdns-publisher` is not run on the bootstrap node, as there is no need for any
other host to discover the IP address that the bootstrap VM gets from DHCP.                                                                                                                                                                             
```

markmc commented on docs/design/baremetal/networking-infrastructure.md:
```diff
@@ -70,7 +70,101 @@ DNS requirements internal to the cluster as possible, leaving only a
+
+The `mdns` plugin for `coredns` was developed to perform DNS lookups
+based on discoverable information from mDNS.  This plugin will resolve both the
+`etcd-NNN` records, as well as the `_etcd-server-ssl._tcp.` SRV record.
```
This begs the question of "what about api-int?" given the "internally resolvable" list above

Maybe move the api-int bit up to come before the mdns bit, especially since it's so simple?

markmc replied:
I mention the master-NN and worker-NN names above in mdns-publisher

So coredns-mdns publishes those too

Why? Should we include those in the list of "internally resolvable DNS records" above?

russellb replied:
I've moved `api-int` up as suggested, rewording it to not mention the "DNS VIP" since the DNS VIP will not have been discussed yet.

I intentionally left out mention of the `master-NN` and `worker-NN` names.  I know they were added because AWS had it, but I actually have no idea what they are used for, or if they're necessary at all.  Looking at the AWS implementation again, I don't think they are even there anymore.

@celebdor @cybertron It looks like we only need etcd records.  We may not need `mdns-publisher` on the workers at all.  Do you know of any reason that is still needed?

cybertron replied:
I don't  know anything specific. The addresses we're serving are the ones that were in the original requirements.

I guess we can always try removing mdns-publisher from the workers and see what breaks.

markmc commented on docs/design/baremetal/networking-infrastructure.md:
```diff
@@ -70,7 +70,101 @@ DNS requirements internal to the cluster as possible, leaving only a
+https://github.com/openshift/mdns-publisher
+
+The `mdns-publisher` is the component that runs on each host to make itself
+discoverable by other hosts in the cluster.  It will be run by the
```
On control plane machines, it advertises using etcd-NN and master-NN names, but on worker nodes only the worker-NN name is published

russellb replied:
Thanks, I've added this, including a note that the master-NN and worker-NN names may no longer be necessary AFAICT ...

markmc commented on docs/design/baremetal/networking-infrastructure.md:
```diff
@@ -70,7 +70,101 @@ DNS requirements internal to the cluster as possible, leaving only a
+control plane nodes currently running.
+
+The IP addresses that the `etcd-NNN` host records resolve to comes from the
+mDNS advertisement sent out by the `mdns-publisher` on that control plane noded
```
Bit of a typo - "that control plane noded"

russellb replied:
fixed

markmc commented:
I think need to cover kubeconfig-extractor (and "clusterrc") somewhere

markmc commented:
This is awesome btw!

russellb committed:
baremetal: Note the network info in the install docs.

russellb commented:
> I think need to cover kubeconfig-extractor (and "clusterrc") somewhere

maybe in a more detailed doc in the MCO repo that discusses those assets?

russellb committed:
baremetal: Add info on using haproxy to load balance API traffic.

russellb committed:
baremetal: Add links to bootstrap asset filenames.

russellb committed:
baremetal: Correct description of VRRP priority usage.

The baremetal VM actually has a higher VRRP priority than the control
plane nodes.  This ensures that the VIP does not move until the API is
fully functional on the control plane nodes.  The signal for this is
when the bootstrap VM is stopped.  That will leave only the lower
priority keepalived instances running that reside on the control plane
nodes.

russellb commented on docs/design/baremetal/networking-infrastructure.md:
```diff
@@ -32,19 +38,36 @@ The API VIP first resides on the bootstrap VM. The `keepalived`
-pods](https://kubernetes.io/docs/tasks/administer-cluster/static-pod/)
-and the relevant assets are [rendered by the Machine Config
+priority. This ensures that the API on the control plane nodes is fully
+functional before the API VIP moves.
```
@markmc you raised the discrepency about VRRP priorities out of band.  I have corrected the documentation to reflect what's happening here.  FYI

markmc commented:
Awesome progress, I think we're saying 2 follows-up needed:

1. need detailed MCO docs - covering dns-vip-prepender and kubeconfig-extractor (and "clusterrc") 
2. look into whether need the master-NNN and worker-NNN records
